qemu/tests/block_transfer_parameters_check.py

```diff
@@ -0,0 +1,155 @@
+    :param env:    Dictionary with test environment.
+    """
+
+    def _verify_transfer_info(host_info, guest_info):
+        errmsg = "The guest get unexpected transfer %s" % guest_info
+        if guest_info["sectors_kb"] > host_info["sectors_kb"]:
+            test.fail(errmsg)
+        min_host = min(host_info["sectors_kb"], host_info["segments"] * 4)
+        logger.debug(min_host)
+        if guest_info["sectors_kb"] > min_host:
+            test.fail(errmsg)
+        if guest_info["tran_length"] > guest_info["sectors_kb"] * 2:
+            test.fail(errmsg)
+
+    def _get_target_devices(trans_type, size=None):
+        devs = []
+        cond = "$2==\"%s\"" % trans_type
+        if size:
+            cond = cond + " && $3==\"%s\"" % size
+        cmd = "lsblk -Spo 'NAME,TRAN,SIZE' |awk '{if(%s) print $1}'" % cond
+        status, output = process.getstatusoutput(cmd)
+        devs_str = output.strip().replace("\n", " ")
+        if devs_str:
+            cmd = "lsblk -Jpo 'NAME,HCTL,SERIAL,TRAN,FSTYPE,WWN' %s" % devs_str
+            status, output = process.getstatusoutput(cmd)
+            devs = copy.deepcopy(json.loads(output)["blockdevices"])
+
+        return devs
+
+    def _run_cmd(cmd, conn=None):
+        if conn:
+            return conn.cmd_output(cmd)
+        else:
+            return process.getoutput(cmd)
+
+    def _get_transfer_parameters(dev, conn=None):
+        dev = basename(dev)
+        info = {"tran_length": 0, "sectors_kb": 0, "segments": 0}
+        cmd = params["get_tran_length_cmd"] % dev
+        output = _run_cmd(cmd, conn)
+        logger.debug(output)
+        info["tran_length"] = int(output)
+        cmd = params["get_tran_params_cmd"] % dev
+        output = _run_cmd(cmd, conn).split()
+        logger.debug(output)
+        info["sectors_kb"] = int(output[0])
+        info["segments"] = int(output[1])
+        return info
+
+    vm = None
+    iscsi = None
+    logger = logging.getLogger('avocado.test')
+    set_max_sector_cmd = params.get("set_max_sector_cmd")
+
+    try:
+        tran_type = params["tran_type"]
+        params['image_size'] = params.get('emulated_image_size', "")
+        if tran_type == "iscsi":
+            logger.debug("Create iscsi disk.")
+            base_dir = data_dir.get_data_dir()
+            iscsi = Iscsi.create_iSCSI(params, base_dir)
+            iscsi.login()
+            dev_name = utils_misc.wait_for(lambda: iscsi.get_device_name(), 60)
+            if not dev_name:
+                test.error('Can not get the iSCSI device.')
+            logger.debug(dev_name)
+            time.sleep(2)
+            logger.debug(_run_cmd("lsblk -JO %s" % dev_name))
+
+        target_devs = _get_target_devices(tran_type, params['image_size'])
+        if not len(target_devs):
+            if tran_type == "fc":
+                test.cancel("No FC device:%s" % params['image_size'])
+            else:
+                test.error("No ISCSI device:%s" % params['image_size'])
```

Comment:
I think we should use test.cancel here, in order to keep the same behavior with fc device. Thanks.

Reply:
FC depends on HW, it can be canceled if no FC on the host. but ISCSI is deployed by code, it could be an error if not deployed succeed

Reply:
OK, thanks.